Pinta.Core/Effects/ColorGradient.cs

```diff
@@ -0,0 +1,183 @@
+		ColorBgra startColor,
+		ColorBgra endColor,
+		double minPosition,
+		double maxPosition,
+		IEnumerable<KeyValuePair<double, ColorBgra>> gradientStops)
+	{
+		CheckBoundsConsistency (minPosition, maxPosition);
+
+		var sortedStops = GetSortedStops (gradientStops);
+		CheckStopsBounds (sortedStops, minPosition, maxPosition);
+		CheckUniqueness (sortedStops);
+
+		StartColor = startColor;
+		EndColor = endColor;
+		MinimumPosition = minPosition;
+		MaximumPosition = maxPosition;
+		sorted_stops = sortedStops;
+	}
+
+	private static IReadOnlyList<GradientStop> GetSortedStops (IEnumerable<KeyValuePair<double, ColorBgra>> gradientStops)
+		=>
+			gradientStops
+			.Select (kvp => new GradientStop (kvp.Key, kvp.Value))
+			.OrderBy (stop => stop.Position)
+			.ToArray ();
+
+	private static void CheckStopsBounds (IReadOnlyList<GradientStop> sortedStops, double minPosition, double maxPosition)
+	{
+		if (sortedStops.Count > 0 && sortedStops[0].Position <= minPosition) throw new ArgumentException ($"Lowest key in gradient stops has to be greater than {nameof (minPosition)}");
+		if (sortedStops.Count > 0 && sortedStops[^1].Position >= maxPosition) throw new ArgumentException ($"Greatest key in gradient stops has to be lower than {nameof (maxPosition)}");
+	}
+
+	private static void CheckUniqueness (IReadOnlyList<GradientStop> sortedStops)
+	{
+		var distinctPositions = sortedStops.GroupBy (s => s.Position).Count ();
+		if (distinctPositions != sortedStops.Count) throw new ArgumentException ("Cannot have more than one stop in the same position");
+	}
+
+	private static void CheckBoundsConsistency (double minPosition, double maxPosition)
+	{
+		if (minPosition >= maxPosition) throw new ArgumentException ($"{nameof (minPosition)} has to be lower than {nameof (maxPosition)}");
+	}
+
+	public ColorGradient Resized (double minPosition, double maxPosition)
+	{
+		CheckBoundsConsistency (minPosition, maxPosition);
+
+		double newSpan = maxPosition - minPosition;
+		double currentSpan = MaximumPosition - MinimumPosition;
+		double newProportion = newSpan / currentSpan;
+		double newMinRelativeOffset = minPosition - MinimumPosition;
+
+		KeyValuePair<double, ColorBgra> ToNewStop (GradientStop stop)
+		{
+			double stopToMinOffset = stop.Position - MinimumPosition;
+			double adjustedOffset = stopToMinOffset * newProportion;
+			double newPosition = minPosition + adjustedOffset;
+			return KeyValuePair.Create (newPosition, stop.Color);
+		}
+
+		return new (
+			StartColor,
+			EndColor,
+			minPosition,
+			maxPosition,
+			sorted_stops.Select (ToNewStop)
+		);
+	}
+
+	public ColorBgra GetColor (double position)
+	{
+		if (position <= MinimumPosition) return StartColor;
+		if (position >= MaximumPosition) return EndColor;
+		if (sorted_stops.Count == 0) return HandleNoStops (position);
+		return HandleWithStops (position);
+	}
+
+	private ColorBgra HandleNoStops (double position)
+	{
+		double valueSpan = MaximumPosition - MinimumPosition;
```

Comment:
This code here could perhaps be factored out a method like `Utility.InvLerp()` and reused in `HandleWithStops` below - it's doing the inverse of `Lerp(MinimumPosition, MaximumPosition, fraction)`

Reply:
I just created and commited an `InvLerp()` method